File: domain/stats/nature.go
package stats

import (
	"damagecalculator/domain/basepoints"
	"damagecalculator/domain/individuals"
)

type NatureType uint

const (
	Bashful NatureType = iota
	Lonely
	Adamant
	Naughty
	Brave
	Bold
	Impish
	Lax
	Relaxed
	Modest
	Mild
	Rash
	Quiet
	Calm
	Gentle
	Careful
	Sassy
	Timid
	Hasty
	Jolly
	Naive
)

type Nature struct {
	hp, at, df, sa, sd, sp statsCalculator
}

type NatureDesc struct {
	Name, Desc string
}

func NatureDescriptions() []NatureDesc {
	res := make([]NatureDesc, 0)
	list := []NatureType{
		Bashful,
		Lonely,
		Adamant,
		Naughty,
		Brave,
		Bold,
		Impish,
		Lax,
		Relaxed,
		Modest,
		Mild,
		Rash,
		Quiet,
		Calm,
		Gentle,
		Careful,
		Sassy,
		Timid,
		Hasty,
		Jolly,
		Naive,
	}
	for _, n := range list {
		name := nameMap[n]
		desc := descMap[n]
		res = append(res, NatureDesc{Name: name, Desc: desc})
	}
	return res
}

func (n *Nature) calcHP(l Level, s Species, i individuals.Individual, b basePoints.BasePoint) uint {
	return n.hp(l, s, i, b)
}
func (n *Nature) calcAttack(l Level, s Species, i individuals.Individual, b basePoints.BasePoint) uint {
	return n.at(l, s, i, b)
}
func (n *Nature) calcDefense(l Level, s Species, i individuals.Individual, b basePoints.BasePoint) uint {
	return n.df(l, s, i, b)
}
func (n *Nature) calcSpAttack(l Level, s Species, i individuals.Individual, b basePoints.BasePoint) uint {
	return n.sa(l, s, i, b)
}
func (n *Nature) calcSpDefense(l Level, s Species, i individuals.Individual, b basePoints.BasePoint) uint {
	return n.sd(l, s, i, b)
}
func (n *Nature) calcSpeed(l Level, s Species, i individuals.Individual, b basePoints.BasePoint) uint {
	return n.sp(l, s, i, b)
}

func NameToNature(name string) *Nature {
	for k, v := range nameMap {
		if v == name {
			return NewNature(k)
		}
	}
	return NewNature(Bashful)
}
func NameToNatureType(name string) NatureType {
	for k, v := range nameMap {
		if v == name {
			return k
		}
	}
	return Bashful
}

func NewNature(n NatureType) *Nature {
	return newNatureFuncs[n]()
}
func (n *Nature) ToShadinja() {
	n.hp = calcShadinjaHP
}

type newNatureFunc func() *Nature

var newNatureFuncs map[NatureType]newNatureFunc

var nameMap map[NatureType]string
var descMap map[NatureType]string

func init() {
	descMap = map[NatureType]string{
		Bashful: "補正なし",
		Lonely:  "攻撃1.1 防御0.9",
		Adamant: "攻撃1.1 特攻0.9",
		Naughty: "攻撃1.1 特防0.9",
		Brave:   "攻撃1.1 素早さ0.9",
		Bold:    "防御1.1 攻撃0.9",
		Impish:  "防御1.1 特攻0.9",
		Lax:     "防御1.1 特防0.9",
		Relaxed: "防御1.1 素早さ0.9",
		Modest:  "特攻1.1 攻撃0.9",
		Mild:    "特攻1.1 防御0.9",
		Rash:    "特攻1.1 特防0.9",
		Quiet:   "特攻1.1 素早さ0.9",
		Calm:    "特防1.1　攻撃0.9",
		Gentle:  "特防1.1 防御0.9",
		Careful: "特防1.1 特攻0.9",
		Sassy:   "特防1.1　素早さ0.9",
		Timid:   "素早さ1.1 攻撃0.9",
		Hasty:   "素早さ1.1 防御0.9",
		Jolly:   "素早さ1.1 特攻0.9",
		Naive:   "素早さ1.1 特防0.9",
	}
	nameMap = map[NatureType]string{
		Bashful: "てれや",
		Lonely:  "さみしがり",
		Adamant: "いじっぱり",
		Naughty: "やんちゃ",
		Brave:   "ゆうかん",
		Bold:    "ずぶとい",
		Impish:  "わんぱく",
		Lax:     "のうてんき",
		Relaxed: "のんき",
		Modest:  "ひかえめ",
		Mild:    "おっとり",
		Rash:    "うっかりや",
		Quiet:   "れいせい",
		Calm:    "おだやか",
		Gentle:  "おとなしい",
		Careful: "しんちょう",
		Sassy:   "なまいき",
		Timid:   "おくびょう",
		Hasty:   "せっかち",
		Jolly:   "ようき",
		Naive:   "むじゃき",
	}
	newNatureFuncs = map[NatureType]newNatureFunc{
		Bashful: bashful,
		Lonely:  lonely,
		Adamant: adamant,
		Naughty: naughty,
		Brave:   brave,
		Bold:    bold,
		Impish:  impish,
		Lax:     lax,
		Relaxed: relaxed,
		Modest:  modest,
		Mild:    mild,
		Rash:    rash,
		Quiet:   quiet,
		Calm:    calm,
		Gentle:  gentle,
		Careful: careful,
		Sassy:   sassy,
		Timid:   timid,
		Hasty:   hasty,
		Jolly:   jolly,
		Naive:   naive,
	}
}

func bashful() *Nature {
	res := new(Nature)
	res.hp = calcHP
	res.at = calcFlat
	res.df = calcFlat
	res.sa = calcFlat
	res.sd = calcFlat
	res.sp = calcFlat
	return res
}
func lonely() *Nature {
	res := bashful()
	res.at = calcUpper
	res.df = calcLower
	return res
}
func adamant() *Nature {
	res := bashful()
	res.at = calcUpper
	res.sa = calcLower
	return res
}
func naughty() *Nature {
	res := bashful()
	res.at = calcUpper
	res.sd = calcLower
	return res
}
func brave() *Nature {
	res := bashful()
	res.at = calcUpper
	res.sp = calcLower
	return res
}
func bold() *Nature {
	res := bashful()
	res.df = calcUpper
	res.at = calcLower
	return res
}
func impish() *Nature {
	res := bashful()
	res.df = calcUpper
	res.sa = calcLower
	return res
}
func lax() *Nature {
	res := bashful()
	res.df = calcUpper
	res.sd = calcLower
	return res
}
func relaxed() *Nature {
	res := bashful()
	res.df = calcUpper
	res.sp = calcLower
	return res
}
func modest() *Nature {
	res := bashful()
	res.sa = calcUpper
	res.at = calcLower
	return res
}
func mild() *Nature {
	res := bashful()
	res.sa = calcUpper
	res.df = calcLower
	return res
}
func rash() *Nature {
	res := bashful()
	res.sa = calcUpper
	res.sd = calcLower
	return res
}
func quiet() *Nature {
	res := bashful()
	res.sa = calcUpper
	res.sp = calcLower
	return res
}
func calm() *Nature {
	res := bashful()
	res.sd = calcUpper
	res.at = calcLower
	return res
}
func gentle() *Nature {
	res := bashful()
	res.sd = calcUpper
	res.df = calcLower
	return res
}
func careful() *Nature {
	res := bashful()
	res.sd = calcUpper
	res.sa = calcLower
	return res
}
func sassy() *Nature {
	res := bashful()
	res.sd = calcUpper
	res.sp = calcLower
	return res
}
func timid() *Nature {
	res := bashful()
	res.sp = calcUpper
	res.at = calcLower
	return res
}
func hasty() *Nature {
	res := bashful()
	res.sp = calcUpper
	res.df = calcLower
	return res
}
func jolly() *Nature {
	res := bashful()
	res.sp = calcUpper
	res.sa = calcLower
	return res
}
func naive() *Nature {
	res := bashful()
	res.sp = calcUpper
	res.sd = calcLower
	return res
}

File: frontend/src/domain/nature.ts
/*
    性格補正
*/
export interface INature {
    Name(): string;
    Attack(v: number): number;
    Defense(v: number): number;
    SpAttack(v: number): number;
    SpDefense(v: number): number;
    Speed(v: number): number;
}

// てれや:補正なし
class Bashful implements INature {
    Name(): string { return 'てれや'; }
    Attack(v: number): number { return v; }
    Defense(v: number): number { return v; }
    SpAttack(v: number): number { return v; }
    SpDefense(v: number): number { return v; }
    Speed(v: number): number { return v; }
}
class Lonely extends Bashful {
    Name(): string { return 'さみしがり'; }
    Attack(v: number): number { return Math.floor(v * 1.1); }
    Defense(v: number): number { return Math.floor(v * 0.9); }
}
class Adamant extends Bashful {
    Name(): string { return 'いじっぱり'; }
    Attack(v: number): number { return Math.floor(v * 1.1); }
    SpAttack(v: number): number { return Math.floor(v * 0.9); }
}
class Naughty extends Bashful {
    Name(): string { return 'やんちゃ'; }
    Attack(v: number): number { return Math.floor(v * 1.1); }
    SpDefense(v: number): number { return Math.floor(v * 0.9); }
}
class Brave extends Bashful {
    Name(): string { return 'ゆうかん'; }
    Attack(v: number): number { return Math.floor(v * 1.1); }
    Speed(v: number): number { return Math.floor(v * 0.9); }
}

class Bold extends Bashful {
    Name(): string { return 'ずぶとい'; }
    Defense(v: number): number { return Math.floor(v * 1.1); }
    Attack(v: number): number { return Math.floor(v * 0.9); }
}
class Impish extends Bashful {
    Name(): string { return 'わんぱく'; }
    Defense(v: number): number { return Math.floor(v * 1.1); }
    SpAttack(v: number): number { return Math.floor(v * 0.9); }
}
class Lax extends Bashful {
    Name(): string { return 'のうてんき'; }
    Defense(v: number): number { return Math.floor(v * 1.1); }
    SpDefense(v: number): number { return Math.floor(v * 0.9); }
}
class Relaxed extends Bashful {
    Name(): string { return 'のんき'; }
    Defense(v: number): number { return Math.floor(v * 1.1); }
    Speed(v: number): number { return Math.floor(v * 0.9); }
}

class Modest extends Bashful {
    Name(): string { return 'ひかえめ'; }
    SpAttack(v: number): number { return Math.floor(v * 1.1); }
    Attack(v: number): number { return Math.floor(v * 0.9); }
}
class Mild extends Bashful {
    Name(): string { return 'おっとり'; }
    SpAttack(v: number): number { return Math.floor(v * 1.1); }
    Defense(v: number): number { return Math.floor(v * 0.9); }
}
class Rash extends Bashful {
    Name(): string { return 'うっかりや'; }
    SpAttack(v: number): number { return Math.floor(v * 1.1); }
    SpDefense(v: number): number { return Math.floor(v * 0.9); }
}
class Quiet extends Bashful {
    Name(): string { return 'れいせい'; }
    SpAttack(v: number): number { return Math.floor(v * 1.1); }
    Speed(v: number): number { return Math.floor(v * 0.9); }
}

class Calm extends Bashful {
    Name(): string { return 'おだやか'; }
    SpDefense(v: number): number { return Math.floor(v * 1.1); }
    Attack(v: number): number { return Math.floor(v * 0.9); }
}
class Gentle extends Bashful {
    Name(): string { return 'おとなしい'; }
    SpDefense(v: number): number { return Math.floor(v * 1.1); }
    Defense(v: number): number { return Math.floor(v * 0.9); }
}
class Careful extends Bashful {
    Name(): string { return 'しんちょう'; }
    SpDefense(v: number): number { return Math.floor(v * 1.1); }
    SpAttack(v: number): number { return Math.floor(v * 0.9); }
}
class Sassy extends Bashful {
    Name(): string { return 'なまいき'; }
    SpDefense(v: number): number { return Math.floor(v * 1.1); }
    Speed(v: number): number { return Math.floor(v * 0.9); }
}

class Timid extends Bashful {
    Name(): string { return 'おくびょう'; }
    Speed(v: number): number { return Math.floor(v * 1.1); }
    Attack(v: number): number { return Math.floor(v * 0.9); }
}
class Hasty extends Bashful {
    Name(): string { return 'せっかち'; }
    Speed(v: number): number { return Math.floor(v * 1.1); }
    Defense(v: number): number { return Math.floor(v * 0.9); }
}
class Jolly extends Bashful {
    Name(): string { return 'ようき'; }
    Speed(v: number): number { return Math.floor(v * 1.1); }
    SpAttack(v: number): number { return Math.floor(v * 0.9); }
}
class Naive extends Bashful {
    Name(): string { return 'むじゃき'; }
    Speed(v: number): number { return Math.floor(v * 1.1); }
    SpDefense(v: number): number { return Math.floor(v * 0.9); }
}

let naturesList: INature[] = [
    new Bashful,
    new Lonely,
    new Adamant,
    new Naughty,
    new Brave,
    new Bold,
    new Impish,
    new Lax,
    new Relaxed,
    new Modest,
    new Mild,
    new Rash,
    new Quiet,
    new Calm,
    new Gentle,
    new Careful,
    new Sassy,
    new Timid,
    new Hasty,
    new Jolly,
    new Naive,
];
export function NatureNames(): string[] {
    let res = naturesList.map(n => n.Name());
    return res;
}

export function NatureFactory(name: string): INature {
    // a! undefinedを許容しない
    let res = naturesList.find(n => n.Name() == name)!;
    return res;
}


// 性格
class nature {
    constructor(at, df, sa, sd, sp, name) {
        this.at = at;
        this.df = df;
        this.sa = sa;
        this.sd = sd;
        this.sp = sp;
        this.name = name;
    }
    attack(val) {
        return Math.floor(val * this.at);
    }
    defense(val) {
        return Math.floor(val * this.df);
    }
    spAttack(val) {
        return Math.floor(val * this.sa);
    }
    spDefense(val) {
        return Math.floor(val * this.sd);
    }
    speed(val) {
        return Math.floor(val * this.sp);
    }
}
// 性格の名前一覧
export function natureNames() {
    return natures.map(n => n.name);
}
// 名前から性格を取得
export function getNature(name) {
    return natures.find(n => n.name == name);
}

let natures = [];
natures.push(new nature(1.0, 1.0, 1.0, 1.0, 1.0, 'てれや'))

natures.push(new nature(1.1, 0.9, 1.0, 1.0, 1.0, 'さみしがり'))
natures.push(new nature(1.1, 1.0, 0.9, 1.0, 1.0, 'いじっぱり'))
natures.push(new nature(1.1, 1.0, 1.0, 0.9, 1.0, 'やんちゃ'))
natures.push(new nature(1.1, 1.0, 1.0, 1.0, 0.9, 'ゆうかん'))

natures.push(new nature(0.9, 1.1, 1.0, 1.0, 1.0, 'ずぶとい'))
natures.push(new nature(1.0, 1.1, 0.9, 1.0, 1.0, 'わんぱく'))
natures.push(new nature(1.0, 1.1, 1.0, 0.9, 1.0, 'のうてんき'))
natures.push(new nature(1.0, 1.1, 1.0, 1.0, 0.9, 'のんき'))

natures.push(new nature(0.9, 1.0, 1.1, 1.0, 1.0, 'ひかえめ'))
natures.push(new nature(1.0, 0.9, 1.1, 1.0, 1.0, 'おっとり'))
natures.push(new nature(1.0, 1.0, 1.1, 0.9, 1.0, 'うっかりや'))
natures.push(new nature(1.0, 1.0, 1.1, 1.0, 0.9, 'れいせい'))

natures.push(new nature(0.9, 1.0, 1.0, 1.1, 1.0, 'おだやか'))
natures.push(new nature(1.0, 0.9, 1.0, 1.1, 1.0, 'おとなしい'))
natures.push(new nature(1.0, 1.0, 0.9, 1.1, 1.0, 'しんちょう'))
natures.push(new nature(1.0, 1.0, 1.0, 1.1, 0.9, 'なまいき'))

natures.push(new nature(0.9, 1.0, 1.0, 1.0, 1.1, 'おくびょう'))
natures.push(new nature(1.0, 0.9, 1.0, 1.0, 1.1, 'せっかち'))
natures.push(new nature(1.0, 1.0, 0.9, 1.0, 1.1, 'ようき'))
natures.push(new nature(1.0, 1.0, 1.0, 0.9, 1.1, 'むじゃき'))
